src/main/scala/analysis/MemoryModelMap.scala

```diff
@@ -4,67 +4,156 @@ import analysis.*
+  private val bufferedSharedStackMap: mutable.Map[String, mutable.Map[Procedure, mutable.TreeMap[RangeKey, StackRegion]]] = mutable.Map()
   private val heapMap: mutable.Map[RangeKey, HeapRegion] = mutable.TreeMap()
   private val dataMap: mutable.Map[RangeKey, DataRegion] = mutable.TreeMap()
+  private val cfgPositionToDataRegion: mutable.Map[CFGPosition, Set[DataRegion]] = mutable.Map()
+  private val heapCalls: mutable.Map[DirectCall, HeapRegion] = mutable.Map()
+  private val mergedRegions: mutable.Map[Set[MemoryRegion], String] = mutable.Map()
+
+  private val stackAllocationSites: mutable.Map[CFGPosition, Set[StackRegion]] = mutable.Map()
+
+  private val uf = new UnionFind()
+  private var DataMemory, HeapMemory, StackMemory = TreeMap[BigInt, Array[Byte]]()
+
+
+
+  // Store operation: store BigInt value at a BigInt address
+  def store(address: BigInt, value: BigInt, memoryType: MemoryType): Unit = {
+    val byteArray = value.toByteArray
+    memoryType match
+      case MemoryType.Data => DataMemory += (address -> byteArray)
+      case MemoryType.Heap => HeapMemory += (address -> byteArray)
+      case MemoryType.Stack => StackMemory += (address -> byteArray)
+  }
+
+  // Load operation: load from a BigInt address with a specific size
+  def load(address: BigInt, size: Int, memoryType: MemoryType): BigInt = {
+    val memory = memoryType match
+      case MemoryType.Data => DataMemory
+      case MemoryType.Heap => HeapMemory
+      case MemoryType.Stack => StackMemory
+    // Find the memory block that contains the starting address
+    val floorEntry = memory.rangeTo(address).lastOption
+
+    floorEntry match {
+      case Some((startAddress, byteArray)) =>
+        val offset = (address - startAddress).toInt // Offset within the byte array
+        // If the load exceeds the stored data, we need to handle padding with zeros
+        if (offset >= byteArray.length) {
+          BigInt(0)
+        } else {
+          // Calculate how much data we can retrieve
+          val availableSize = byteArray.length - offset
+          // Slice the available data, and if requested size exceeds, append zeros
+          val result = byteArray.slice(offset, offset + size)
+          val paddedResult = if (size > availableSize) {
+            result ++ Array.fill(size - availableSize)(0.toByte) // Padding with zeros
+          } else {
+            result
+          }
+          BigInt(1, paddedResult) // Convert the byte array back to BigInt
+        }
+      case None =>
+        // If no memory is stored at the requested address, return zero
+        BigInt(0) // TODO: may need to be sm else
+    }
+  }
 
   /** Add a range and object to the mapping
    *
-   * @param offset the offset of the range
+   * @param offset the offset of the range, if a heap region is given, the offsets controls the shift of regions from the start
```

Comment:
What does this mean?

Reply:
It is a way to give heap regions an assumed "offset". This should mean the offset passed is alway 0 but not explained properly. I will modify the function and the comments to be clearer

Reply:
Why do you give heap regions an assumed offset at all if you always use 0?